File: netlify/functions/collect-images.js
import crypto from 'crypto'
import { requireAuth, adminSupabase, ok, err, options, isValidUUID } from './utils/supabase.js'
import { getUserUsage } from './utils/usage.js'

const PLATFORM_KEY = process.env.GOOGLE_MAPS_KEY
const CAP          = 20

// Resolves Google API key(s) to use for billing and whether to deduct purchased credits.
// Admin           → platform key for everything, no credit deduction.
// Non-admin       → ALWAYS deducts purchased credits (lifetime balance, regardless of key).
//   No purchased credits remaining → blocked.
//   Own key covers up to (points_limit - used) downloads remaining this cycle;
//   any points beyond that — even within the same batch — fall back to the
//   platform key, so a batch straddling the monthly boundary is split correctly.
async function resolveApiKeyAndMode(userId, supabase) {
  const [{ data: keyRow }, { data: profile }, usage] = await Promise.all([
    supabase.from('user_keys').select('google_maps_key').eq('user_id', userId).maybeSingle(),
    supabase.from('profiles').select('role').eq('id', userId).maybeSingle(),
    getUserUsage(userId, supabase),
  ])

  if (profile?.role === 'admin') {
    return { ownKey: null, ownKeyCapacity: 0, platformKey: PLATFORM_KEY, deductPurchased: false, remaining: Infinity }
  }

  const { used, limit, purchasedRemaining } = usage

  if (purchasedRemaining <= 0) {
    return { ownKey: null, ownKeyCapacity: 0, platformKey: null, deductPurchased: false, remaining: 0 }
  }

  return {
    ownKey:          keyRow?.google_maps_key ?? null,
    ownKeyCapacity:  keyRow?.google_maps_key ? Math.max(0, limit - used) : 0,
    platformKey:     PLATFORM_KEY,
    deductPurchased: true,
    remaining:       purchasedRemaining,
  }
}

// Fetch the actual Street View panorama location (free metadata, no charge).
// Returns { lat, lng } of where Google's camera physically is, or null.
async function getPanoramaLocation(lat, lng, apiKey) {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), 5000)
  try {
    const res  = await fetch(
      `https://maps.googleapis.com/maps/api/streetview/metadata?location=${lat},${lng}&return_error_code=true&key=${apiKey}`,
      { signal: controller.signal }
    )
    const data = await res.json()
    if (data.status === 'OK' && data.location) return data.location
    if (data.status && data.status !== 'ZERO_RESULTS' && data.status !== 'NOT_FOUND') {
      console.error(`Street View metadata rejected: ${data.status}${data.error_message ? ' — ' + data.error_message : ''}`)
    }
  } catch { /* timeout or network error — fall back to road_bearing */ }
  finally { clearTimeout(timer) }
  return null
}

// Compass bearing (degrees, 0=North) from one coordinate to another.
function bearingTo(fromLat, fromLng, toLat, toLng) {
  const lat1 = fromLat * Math.PI / 180
  const lat2 = toLat   * Math.PI / 180
  const dLng = (toLng - fromLng) * Math.PI / 180
  const y = Math.sin(dLng) * Math.cos(lat2)
  const x = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLng)
  return (Math.atan2(y, x) * 180 / Math.PI + 360) % 360
}

// Returns: { buffer } on success, { noCoverage: true } for 404 (no imagery),
// or throws an Error for 400/403 (bad key / API not enabled).
async function downloadGoogleImage(lat, lng, heading, apiKey) {
  const url = `https://maps.googleapis.com/maps/api/streetview?size=640x640&location=${lat},${lng}&heading=${heading}&pitch=0&fov=60&return_error_code=true&key=${apiKey}`
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), 5000)
  let res
  try {
    res = await fetch(url, { signal: controller.signal })
  } finally {
    clearTimeout(timer)
  }
  if (res.status === 403 || res.status === 400) {
    // Google returns a plain-text reason (e.g. referrer/IP restriction, API not
    // enabled, billing disabled) in the body — log it so the real cause is
    // visible instead of just "HTTP 403".
    const reason = await res.text().catch(() => '')
    console.error(`Street View download rejected (HTTP ${res.status}):`, reason.slice(0, 500))
    throw new Error(`Street View API key error (HTTP ${res.status}) — check the key is valid and Street View Static API is enabled.`)
  }
  if (!res.ok) return { noCoverage: true }
  if (!(res.headers.get('content-type') || '').includes('image')) return { noCoverage: true }
  return { buffer: await res.arrayBuffer() }
}

async function processPoint(pt, projectId, userId, apiKey, supabase) {
  const { id: pointId, lat, lng, road_bearing, property_lat, property_lng } = pt

  try {
    if (!apiKey) return { pointId, status: 'error', error: 'No Google Maps API key configured' }

    // Aim the camera at the exact point geocode-points.js resolved the address
    // from (property_lat/lng), when it's known — not the raw road-snapped scan
    // point. Using two independently-guessed offsets for the photo vs. the
    // address let them drift onto different houses on curves, corner lots, or
    // uneven parcel spacing; sharing one target point keeps them in sync.
    // Falls back to the raw scan point if geocoding hasn't run yet or never
    // resolved a house-numbered address.
    const targetLat = property_lat ?? lat
    const targetLng = property_lng ?? lng

    // Get the actual panorama position (free metadata call) and aim the camera
    // from there toward the target point. This always faces the property regardless
    // of which side of the road it's on or whether road_bearing is available.
    // Falls back to road_bearing + 90 only if the metadata call fails.
    const pano = await getPanoramaLocation(lat, lng, apiKey)
    let heading
    if (pano) {
      const dist = Math.abs(pano.lat - targetLat) + Math.abs(pano.lng - targetLng)
      heading = dist > 1e-7
        ? Math.round(bearingTo(pano.lat, pano.lng, targetLat, targetLng))   // panorama → property point
        : Math.round((road_bearing ?? 0) + 90) % 360                       // same spot, fall back
    } else {
      heading = Math.round((road_bearing ?? 0) + 90) % 360
    }

    await supabase.from('scan_points')
      .update({ status: 'downloading', updated_at: new Date().toISOString() })
      .eq('id', pointId)

    const result = await downloadGoogleImage(lat, lng, heading, apiKey)

    if (result.noCoverage) {
      await supabase.from('scan_points')
        .update({ status: 'no_coverage', updated_at: new Date().toISOString() })
        .eq('id', pointId)
      return { pointId, status: 'no_coverage' }
    }

    const { buffer } = result

    const storagePath = `${projectId}/${pointId}/F.jpg`
    const { error: upErr } = await supabase.storage
      .from('street-view-images')
      .upload(storagePath, buffer, { contentType: 'image/jpeg', upsert: true })

    if (upErr) {
      await supabase.from('scan_points')
        .update({ status: 'failed', error_msg: upErr.message, updated_at: new Date().toISOString() })
        .eq('id', pointId)
      return { pointId, status: 'failed' }
    }

    const { data: { publicUrl } } = supabase.storage.from('street-view-images').getPublicUrl(storagePath)

    await supabase.from('images').insert({
      scan_point_id: pointId,
      direction:     'F',
      heading,
      storage_path:  storagePath,
      storage_url:   publicUrl,
      image_hash:    crypto.createHash('sha256').update(Buffer.from(buffer)).digest('hex'),
      image_source:  'google',
      size_bytes:    buffer.byteLength,
    })

    await supabase.from('scan_points')
      .update({ status: 'downloaded', error_msg: null, updated_at: new Date().toISOString() })
      .eq('id', pointId)

    await supabase.from('usage_logs').insert({
      user_id:  userId,
      service:  'street_view',
      action:   'image_download',
      count:    1,
      cost_usd: 0.007,
      metadata: { projectId, pointId },
    })

    return { pointId, status: 'downloaded' }
  } catch (e) {
    console.error(`processPoint failed ${pointId}:`, e.message)
    // Re-throw API key errors so the handler can return a 503 to the frontend
    if (e.message.includes('API key error')) throw e
    await supabase.from('scan_points')
      .update({ status: 'failed', error_msg: e.message, updated_at: new Date().toISOString() })
      .eq('id', pointId)
    return { pointId, status: 'failed' }
  }
}

export const handler = async (event) => {
  if (event.httpMethod === 'OPTIONS') return options()
  if (event.httpMethod !== 'POST') return err('Method not allowed', 405)

  const { user, error } = await requireAuth(event)
  if (error) return err(error, 401)

  let projectId, pointIds
  try { ({ projectId, pointIds } = JSON.parse(event.body || '{}')) }
  catch { return err('Invalid request body', 400) }
  if (!isValidUUID(projectId) || !Array.isArray(pointIds) || !pointIds.length) {
    return err('projectId and pointIds required')
  }
  const validIds = pointIds.filter(isValidUUID)
  if (!validIds.length) return err('No valid pointIds')

  const supabase = adminSupabase()

  // Verify project belongs to this user
  const { data: project } = await supabase
    .from('projects').select('id, point_spacing_meters').eq('id', projectId).eq('user_id', user.id).maybeSingle()
  if (!project) return err('Project not found', 404)

  const { ownKey, ownKeyCapacity, platformKey, deductPurchased, remaining } = await resolveApiKeyAndMode(user.id, supabase)
  if (!ownKey && !platformKey) return err('No API key or credits available. Contact your admin.', 503)
  if (remaining <= 0) return err('Insufficient credits — contact your admin to add more credits.', 429)

  const ids = validIds.slice(0, Math.min(CAP, remaining === Infinity ? CAP : remaining))

  const { data: pts } = await supabase
    .from('scan_points')
    .select('id, lat, lng, road_bearing, property_lat, property_lng')
    .in('id', ids)

  if (!pts?.length) return ok({ results: [] })

  // Proximity dedup: only download one image per property.
  // Check against already-downloaded points in DB (cross-batch) and within this batch.
  // Duplicate scan points get the source's image copied — no extra download, but still
  // have an image record so AI analysis can run on them.
  // Use half the spacing as dedup radius — in urban areas full spacing would incorrectly
  // treat two neighboring properties as duplicates.
  const spacingDeg = ((project.point_spacing_meters || 30) / 2) / 111320
  const minLat = Math.min(...pts.map(p => p.lat)) - spacingDeg
  const maxLat = Math.max(...pts.map(p => p.lat)) + spacingDeg
  const minLng = Math.min(...pts.map(p => p.lng)) - spacingDeg
  const maxLng = Math.max(...pts.map(p => p.lng)) + spacingDeg
  const { data: donePts } = await supabase
    .from('scan_points').select('id, lat, lng')
    .eq('project_id', projectId)
    .gte('lat', minLat).lte('lat', maxLat)
    .gte('lng', minLng).lte('lng', maxLng)
    .in('status', ['downloaded', 'complete'])

  const primaries  = []                 // pts to actually download
  const duplicates = []                 // { pt, sourceId } — copy image from sourceId
  for (const pt of pts) {
    const dbDone = (donePts || []).find(d =>
      Math.abs(d.lat - pt.lat) <= spacingDeg && Math.abs(d.lng - pt.lng) <= spacingDeg
    )
    if (dbDone) { duplicates.push({ pt, sourceId: dbDone.id }); continue }

    const batchPrimary = primaries.find(p =>
      Math.abs(p.lat - pt.lat) <= spacingDeg && Math.abs(p.lng - pt.lng) <= spacingDeg
    )
    if (batchPrimary) duplicates.push({ pt, sourceId: batchPrimary.id })
    else              primaries.push(pt)
  }

  const settled = await Promise.allSettled(
    primaries.map((pt, i) => processPoint(pt, projectId, user.id, i < ownKeyCapacity ? ownKey : platformKey, supabase))
  )

  // Surface API key errors as 503 so the frontend scan aborts with a clear message
  const keyError = settled.find(s => s.status === 'rejected' && s.reason?.message?.includes('API key error'))
  if (keyError) return err(keyError.reason.message, 503)

  // Copy the source image to each duplicate scan point
  const dupResults = await Promise.allSettled(
    duplicates.map(async ({ pt, sourceId }) => {
      const { data: src } = await supabase.from('images')
        .select('direction, heading, storage_path, storage_url, image_hash, image_source, size_bytes')
        .eq('scan_point_id', sourceId).limit(1).maybeSingle()
      if (src) {
        await supabase.from('images').insert({ scan_point_id: pt.id, ...src })
        await supabase.from('scan_points')
          .update({ status: 'downloaded', updated_at: new Date().toISOString() })
          .eq('id', pt.id)
        return { pointId: pt.id, status: 'downloaded' }
      }
      // Source had no image (no_coverage/failed) — match that status
      const { data: srcPt } = await supabase.from('scan_points')
        .select('status').eq('id', sourceId).maybeSingle()
      const status = srcPt?.status === 'no_coverage' ? 'no_coverage' : 'complete'
      await supabase.from('scan_points')
        .update({ status, updated_at: new Date().toISOString() }).eq('id', pt.id)
      return { pointId: pt.id, status }
    })
  )

  const results = [
    ...settled.map(s => s.status === 'fulfilled' ? s.value : { pointId: null, status: 'error' }),
    ...dupResults.map(s => s.status === 'fulfilled' ? s.value : { pointId: null, status: 'error' }),
  ]

  // Non-admin users always deduct from their lifetime purchased/granted credit
  // balance, regardless of which key was billed (deductPurchased is false only
  // for admins — see resolveApiKeyAndMode).
  // Only count real downloads (primaries) — copied duplicate images cost nothing.
  const downloadedCount = settled.filter(s => s.status === 'fulfilled' && s.value?.status === 'downloaded').length
  if (deductPurchased && downloadedCount > 0) {
    await supabase.rpc('increment_purchased_credits_used', {
      p_user_id: user.id,
      p_points:  downloadedCount,
    })
  }

  const { count: completed } = await supabase
    .from('scan_points')
    .select('*', { count: 'exact', head: true })
    .eq('project_id', projectId)
    .in('status', ['downloaded', 'analyzing', 'complete'])

  await supabase.from('projects').update({
    completed_points: completed || 0,
    status:           'collecting',
    updated_at:       new Date().toISOString(),
  }).eq('id', projectId)

  return ok({ results })
}

File: netlify/functions/collect-images/collect-images.py
"""
collect-images — Python Netlify function

Priority order for image sourcing:
  1. streetlevel GSV (free, downloads + crops panorama)
  2. Mapillary (free, fast direct download)
  3. no_coverage if both fail
"""

import io
import json
import math
import hashlib
import os
import statistics
from datetime import datetime, timezone
from urllib.request import urlopen, Request

from PIL import Image
from supabase import create_client
from streetlevel import streetview

# ── Config ────────────────────────────────────────────────────────────────────

MAPILLARY_KEY        = os.environ.get('MAPILLARY_ACCESS_TOKEN', '')
SUPABASE_URL         = os.environ['VITE_SUPABASE_URL']
SUPABASE_SERVICE_KEY = os.environ['SUPABASE_SERVICE_ROLE_KEY']

MAPILLARY_RADIUS_M    = 50
PERPENDICULAR_TOL_DEG = 65
CAP                   = 20   # max points per function call

CORS = {
    'Access-Control-Allow-Origin':  '*',
    'Access-Control-Allow-Headers': 'Content-Type, Authorization',
    'Content-Type':                 'application/json',
}

# ── Helpers ───────────────────────────────────────────────────────────────────

def _ok(body):
    return {'statusCode': 200, 'headers': CORS, 'body': json.dumps(body)}

def _err(msg, code=400):
    return {'statusCode': code, 'headers': CORS, 'body': json.dumps({'error': msg})}

def _angular_diff(a, b):
    d = abs(((a - b) % 360 + 360) % 360)
    return 360 - d if d > 180 else d

def _fetch_json(url, headers=None):
    req = Request(url, headers=headers or {})
    with urlopen(req, timeout=10) as res:
        return json.loads(res.read())

def _download_bytes(url, timeout=15):
    """Return (bytes, content_type)."""
    with urlopen(url, timeout=timeout) as res:
        return res.read(), res.headers.get('Content-Type', '')

def _crop_equirectangular(img: Image.Image, heading_deg: float, fov_deg: float = 90) -> Image.Image:
    """
    Crop an equirectangular panorama at the given compass heading.
    Returns a 640×480 JPEG-ready RGB image.
    """
    w, h = img.size
    heading_deg = heading_deg % 360

    center_col = int((heading_deg / 360.0) * w)
    crop_w     = int((fov_deg   / 360.0) * w)

    # Use the middle vertical third to avoid sky and road surface
    top    = h // 4
    crop_h = h // 2

    left  = center_col - crop_w // 2
    right = left + crop_w

    if 0 <= left and right <= w:
        cropped = img.crop((left, top, right, top + crop_h))
    elif left < 0:
        p1 = img.crop((w + left, top, w,     top + crop_h))
        p2 = img.crop((0,        top, right, top + crop_h))
        cropped = Image.new('RGB', (crop_w, crop_h))
        cropped.paste(p1, (0, 0))
        cropped.paste(p2, (p1.width, 0))
    else:
        p1 = img.crop((left, top, w,          top + crop_h))
        p2 = img.crop((0,    top, right - w,  top + crop_h))
        cropped = Image.new('RGB', (crop_w, crop_h))
        cropped.paste(p1, (0, 0))
        cropped.paste(p2, (p1.width, 0))

    return cropped.resize((640, 480), Image.LANCZOS)

# ── Mapillary ─────────────────────────────────────────────────────────────────

def _fetch_mapillary(lat, lng):
    """
    Attempt to get a perpendicular image from Mapillary.

    Returns:
        (img_bytes, heading_deg, pano_id, road_bearing_deg)   – on success
        (None,      None,        None,    road_bearing_deg)   – off-perpendicular; road bearing usable
        (None,      None,        None,    None)               – no key or no nearby images
    """
    if not MAPILLARY_KEY:
        return None, None, None, None

    d    = MAPILLARY_RADIUS_M / 111320
    bbox = f'{lng-d},{lat-d},{lng+d},{lat+d}'
    url  = (
        f'https://graph.mapillary.com/images'
        f'?fields=id,thumb_2048_url,compass_angle,computed_compass_angle,geometry,is_pano'
        f'&bbox={bbox}&limit=15'
    )

    try:
        data = _fetch_json(url, headers={'Authorization': f'OAuth {MAPILLARY_KEY}'})
    except Exception as e:
        print(f'[mapillary] API error: {e}')
        return None, None, None, None

    images = data.get('data', [])
    if not images:
        return None, None, None, None

    candidates = []
    for img in images:
        coords = img.get('geometry', {}).get('coordinates', [0, 0])
        dist   = math.hypot((coords[1] - lat) * 111320, (coords[0] - lng) * 111320)
        angle  = img.get('computed_compass_angle') or img.get('compass_angle') or 0
        if img.get('thumb_2048_url'):
            candidates.append({
                'id':     img['id'],
                'url':    img['thumb_2048_url'],
                'angle':  angle,
                'dist':   dist,
                'is_pano': img.get('is_pano', False),
            })

    if not candidates:
        return None, None, None, None

    road_bearing = statistics.median([c['angle'] for c in candidates])
    target_angle = (road_bearing + 90) % 360

    candidates.sort(key=lambda c: (
        (-90 if c['is_pano'] else 0) + _angular_diff(c['angle'], target_angle),
        c['dist'],
    ))

    best      = candidates[0]
    best_diff = _angular_diff(best['angle'], target_angle)

    if not best['is_pano'] and best_diff > PERPENDICULAR_TOL_DEG:
        print(f'[mapillary] best off-perpendicular by {best_diff:.0f}° — skipping')
        return None, None, None, road_bearing

    try:
        img_bytes, ct = _download_bytes(best['url'])
        if 'image' not in ct:
            return None, None, None, road_bearing
        return img_bytes, best['angle'], best['id'], road_bearing
    except Exception as e:
        print(f'[mapillary] download failed: {e}')
        return None, None, None, road_bearing

# ── streetlevel (Google Street View, no API key) ─────────────────────────────

def _fetch_streetlevel(lat, lng, road_bearing=None):
    """
    Download a Google Street View panorama via streetlevel (no API key required),
    crop it perpendicular to the road, and return JPEG bytes.

    Returns:
        (img_bytes, perp_heading_deg, pano_id)  – on success
        None                                     – no panorama found or error
    """
    try:
        pano = streetview.find_panorama(lat, lng, radius=MAPILLARY_RADIUS_M)
        if not pano:
            print(f'[streetlevel] no panorama near {lat},{lng}')
            return None

        # pano.heading is in radians; convert to degrees
        capture_heading_deg = math.degrees(pano.heading) % 360
        base_heading        = road_bearing if road_bearing is not None else capture_heading_deg
        perp_heading        = (base_heading + 90) % 360

        # zoom=1 gives a small but usable resolution — faster than zoom=5
        pano_img = streetview.get_panorama(pano, zoom=1)
        cropped  = _crop_equirectangular(pano_img, perp_heading)

        buf = io.BytesIO()
        cropped.save(buf, format='JPEG', quality=85)
        return buf.getvalue(), perp_heading, pano.id

    except Exception as e:
        print(f'[streetlevel] failed at {lat},{lng}: {e}')
        return None

# ── Per-point pipeline ────────────────────────────────────────────────────────

def _process_point(pt, project_id, user_id, supabase):
    point_id = pt['id']
    lat      = pt['lat']
    lng      = pt['lng']

    try:
        # ── 1. streetlevel GSV ───────────────────────────────────────────
        result = _fetch_streetlevel(lat, lng)
        if result:
            img_bytes, heading, pano_id = result
            img_source = 'google'
        else:
            img_bytes = None

        # ── 2. Mapillary fallback ─────────────────────────────────────────
        if not img_bytes:
            img_bytes, heading, pano_id, _ = _fetch_mapillary(lat, lng)
            img_source = 'mapillary' if img_bytes else None

        if not img_bytes:
            supabase.table('scan_points').update({
                'status':     'no_coverage',
                'updated_at': datetime.now(timezone.utc).isoformat(),
            }).eq('id', point_id).execute()
            return {'pointId': point_id, 'status': 'no_coverage'}

        supabase.table('scan_points').update({
            'status':     'downloading',
            'updated_at': datetime.now(timezone.utc).isoformat(),
        }).eq('id', point_id).execute()

        image_hash   = hashlib.sha256(img_bytes).hexdigest()
        storage_path = f'{project_id}/{point_id}/F.jpg'

        supabase.storage.from_('street-view-images').upload(
            storage_path,
            img_bytes,
            {'content-type': 'image/jpeg', 'upsert': 'true'},
        )

        # Construct public URL directly (avoids an extra round-trip)
        public_url = f'{SUPABASE_URL}/storage/v1/object/public/street-view-images/{storage_path}'

        supabase.table('images').insert({
            'scan_point_id': point_id,
            'direction':     'F',
            'heading':       heading,
            'storage_path':  storage_path,
            'storage_url':   public_url,
            'panorama_id':   pano_id,
            'image_hash':    image_hash,
            'image_source':  img_source,
            'size_bytes':    len(img_bytes),
        }).execute()

        supabase.table('scan_points').update({
            'status':     'downloaded',
            'error_msg':  None,
            'updated_at': datetime.now(timezone.utc).isoformat(),
        }).eq('id', point_id).execute()

        # Both streetlevel GSV and Mapillary are free
        supabase.table('usage_logs').insert({
            'user_id':  user_id,
            'service':  'streetlevel_gsv' if img_source == 'google' else 'mapillary',
            'action':   'image_download',
            'count':    1,
            'cost_usd': 0,
            'metadata': {'projectId': project_id, 'pointId': point_id, 'source': img_source},
        }).execute()

        return {'pointId': point_id, 'status': 'downloaded', 'source': img_source}

    except Exception as e:
        print(f'[collect] error at {point_id}: {e}')
        supabase.table('scan_points').update({
            'status':     'failed',
            'error_msg':  str(e)[:500],
            'updated_at': datetime.now(timezone.utc).isoformat(),
        }).eq('id', point_id).execute()
        return {'pointId': point_id, 'status': 'failed', 'error': str(e)}

# ── Handler ───────────────────────────────────────────────────────────────────

def handler(event, context):
    if event.get('httpMethod') == 'OPTIONS':
        return {'statusCode': 204, 'headers': CORS}

    if event.get('httpMethod') != 'POST':
        return _err('Method not allowed', 405)

    # Auth — verify Supabase JWT via service role
    headers = event.get('headers', {})
    auth    = headers.get('authorization') or headers.get('Authorization', '')
    token   = auth.replace('Bearer ', '').strip()
    if not token:
        return _err('Unauthorized', 401)

    supabase = create_client(SUPABASE_URL, SUPABASE_SERVICE_KEY)
    try:
        user = supabase.auth.get_user(token).user
        if not user:
            return _err('Invalid token', 401)
    except Exception:
        return _err('Invalid token', 401)

    body      = json.loads(event.get('body') or '{}')
    project_id = body.get('projectId')
    point_ids  = body.get('pointIds', [])

    if not project_id or not isinstance(point_ids, list) or not point_ids:
        return _err('projectId and pointIds required')

    ids = point_ids[:CAP]

    pts = (supabase.table('scan_points')
           .select('id, lat, lng, project_id')
           .in_('id', ids)
           .execute()
           .data or [])

    if not pts:
        return _ok({'results': []})

    results = [_process_point(pt, project_id, user.id, supabase) for pt in pts]

    # Update project counters
    completed = (supabase.table('scan_points')
                 .select('id', count='exact')
                 .eq('project_id', project_id)
                 .in_('status', ['downloaded', 'analyzing', 'complete'])
                 .execute()
                 .count or 0)

    supabase.table('projects').update({
        'completed_points': completed,
        'status':           'collecting',
        'updated_at':       datetime.now(timezone.utc).isoformat(),
    }).eq('id', project_id).execute()

    return _ok({'results': results})
